### PR TITLE
Add search handler

### LINKED ISSUE #13391: Add search handlers to allow search logging
### Is your feature request related to problem?
I need a way to be able to log search attempts. Like if a search is leading to 0 hits, I want to know about it and report on it. I think this is something we can easily add by adding handler that is called after a search is executed. 

### Describe the solution you'd like
Add `ISearchHandler` which is called from the search controller. This handler will have Searched and Searching method that would be triggered before and after a search is executed.

### PR DESCRIPTION
**Pull request by MikeAlhayek**
Fix #13391

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 1]
```diff
@@ -142,6 +150,15 @@ public async Task<IActionResult> Search(SearchViewModel viewModel, PagerSlimPara
                 .Where(x => x.ContentItemId.IsIn(searchResult.ContentItemIds) && x.Published);
         }
 
+        var searchContext = new SearchContext()
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:153]:
```suggestion
        var searchContext = new SearchContext
```

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 2]
```diff
@@ -142,6 +150,15 @@ public async Task<IActionResult> Search(SearchViewModel viewModel, PagerSlimPara
...
+        {
+            Index = viewModel.Index,
+            Terms = viewModel.Terms,
+            Hits = await query.CountAsync(),
+        };
+
+        await _searchHandlers.InvokeAsync((handler) => handler.SearchedAsync(searchContext), _logger);
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:160]:
```suggestion
        await _searchHandlers.InvokeAsync(async handler => await handler.SearchedAsync(searchContext), _logger);
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Review by hishamco — APPROVED** [review]
LGTM, @Skrypt let me know if you have any input to this one

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 3]
```diff
@@ -142,6 +150,15 @@ public async Task<IActionResult> Search(SearchViewModel viewModel, PagerSlimPara
...
         }
 
+        var searchContext = new SearchContext
+        {
+            Index = viewModel.Index,
+            Terms = viewModel.Terms,
+            Hits = await query.CountAsync(),
+        };
+
+        await _searchHandlers.InvokeAsync(async handler => await handler.SearchedAsync(searchContext), _logger);
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:160]: @hishamco I saw that you suggested the opposite but here we don't need the async/await. When an async method / func returns another async method, we only need to use async/await if before returning we already needed to call another async method and then already needed to use the await keyword. Both are right but returning the Task directly is less resource consuming, and that's what we are using everywhere else, so I would first suggest.

    await _searchHandlers.InvokeAsync(handler => handler.SearchedAsync(searchContext), _logger);
 
But for optimization `InvokeAsync()` has different overloads using different typed delegates with 0, 1, ... up to 5 generic parameters, then from memory (need to be tried) actual params would need to be passed like this.

    await _searchHandlers.InvokeAsync((handler, context) =>
        handler.SearchedAsync(context), searchContext, _logger);

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:160]: > Both are right but returning the Task directly is less resource consuming

I didn't measure this before, but I TRUST you :)

Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:160]: Okay, but mainly the params should be passed differently, see how `InvokeAsync()` is used in other places.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs]: @jtkech yes I changed it and passed the context as a second parameter. Please let me know if you need any other changes or approve.

**Comment by Skrypt** [comment]
I think it would need to also have a SearchProvider name associated with it. Else, nothing tells which provider the results came from.

**Comment by MikeAlhayek** [comment]
> I think it would need to also have a SearchProvider name associated with it. Else, nothing tells which provider the results came from.

@Skrypt we no longer have or use ` SearchProvider`. We now only use SearchService. I will add `ISearchService` to the context

**Comment by MikeAlhayek** [comment]
@Skrypt I added the `ISearchService` to the SearchContext. Let me know if you like anything else added or please approve this PR.

**Review by Skrypt — APPROVED** [review]

**Pull request merged**